 <a href="https://www.fiap.com.br"><img src = "https://www.fiap.com.br/wp-content/themes/fiap2016/images/sharing/fiap.png" width = 300, align = "center"></a>

<h1 align=center><font size = 5>Análise de dados com Python</font></h1>

# Introdução

### Usando um notebook em Python 

Vamos a abordar a aquisição de dados de várias maneiras e obter as informações necessárias de um conjunto de dados. No final desta aula, você irá carregar / ler dados com sucesso no Jupyter Notebook e obter a descrição da ideia do conjunto de dados através da Biblioteca Pandas. 



## Conteúdo

<div class="alert alert-block alert-info" style="margin-top: 20px">
<li><a href="#ref1">Aquisição de dados</a>
<li><a href="#ref2">Insights Básicos sobre a base de dados</a></li>
<p></p>

</div>
<hr>

<a id="ref1"></a>
# Aquisição de Dados

Existem vários formatos para um conjunto de dados: .csv, .json, .xlsx, etc. Os conjuntos de dados podem ser armazenados em diferentes locais, em sua máquina local ou online.

Nesta seção, você aprenderá como carregar / ler conjunto de dados no nosso Notebook Jupyter.

No nosso caso, o Automobile Dataset é uma fonte on-line, e está no formato CSV (valores separados por vírgula). Vamos usar este conjunto de dados como um exemplo para praticar a leitura de dados.

fonte de dados: https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Tipo de dados : csv

A Biblioteca Pandas é uma ferramenta útil que nos permite ler vários conjuntos de dados em um quadro de dados; a maioria das plataformas de notebooks do Jupyter possui uma Biblioteca de **Pandas** integrada para que tudo o que precisamos fazer é importá-la ou instala-la.


In [1]:
# importe a biblioteca pandas 
import pandas as pd


## Lendo os dados

Usamos o método "pandas.read_csv ()" para ler o arquivo csv. Entre aspas, colocamos o caminho do arquivo junto com uma aspa, de modo que os pandas lerão o arquivo em um quadro de dados desse endereço. O caminho do arquivo pode ser um URL ou o endereço de arquivo local. Como os dados da linha não incluem cabeçalhos, podemos adicionar um argumento **"headers = None"** dentro do método **"read_csv ()"**, de modo que os pandas não definam automaticamente a primeira linha como um cabeçalho. Você também pode atribuir o conjunto de dados a qualquer variável que você criou.


In [39]:

# Importe a pandas como um objeto
import pandas as pd

# leia o arquivo online pelos URLs fornecidos acima e atribua-o a variável "df"
path="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data"

df = pd.read_csv(path,header=None)


In [40]:
df.head()

0    1            2    3    4     5   ...    20   21    22  23  24     25
0   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  13495
1   3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
2   1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
3   2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
4   2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450

[5 rows x 26 columns]

In [37]:
df.head(5)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]



Depois de ler o conjunto de dados, podemos usar o método **`dataframe.head (n)`** para verificar as principais n linhas do quadro de dados onde n é um número inteiro. Confirmar para **`dataframe.head (n)`**, **`dataframe.tail (n)`** irá mostrar-lhe o n linhas inferiores de quadro de dados.


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Pergunta #1: </h1>
<b> Verifique as 10 linhas inferiores do quadro de dados "df".</b>
</div>

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Pergunta #1 Resposta: </h1>
<b>Execute o código abaixo! Você obteve o código certo?</b>
</div>

### Solução

In [ ]:
df.tail(10)

## Adicione os cabeçalhos

Dê uma olhada em nosso conjunto de dados; pandas define automaticamente o cabeçalho por um número inteiro a partir de 0.
<div>
    
Para melhor descrever nossos dados, podemos introduzir um cabeçalho, esta informação está disponível em: https://archive.ics.uci.edu/ml/datasets/Automobile</div>
<p></p>
<div>Assim, temos que adicionar cabeçalhos manualmente.</div>

<div>Primeiramente, nós criamos uma lista com os "headers" que inclui todos os nomes das colunas em ordem.</div>

<div>Então, nós usamos **`dataframe.columns = headers`** para substituir os cabeçalhos a partir da lista que criamos.</div>




In [41]:
# criar lista de cabeçalhos
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]

In [43]:
# substituir cabeçalhos
df.columns = headers

# reveja seu data frame
df.head(7)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450
5          2                 ?         audi  ...       19          25  15250
6          1               158         audi  ...       19          25  17710

[7 rows x 26 columns]

Agora, lemos com sucesso o conjunto de dados bruto e adicionamos os cabeçalhos corretos no quadro de dados.

 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1>Pergunta #2: </h1>
<b>Encontre o nome das colunas do dataframe</b>
</div>

### Solução

In [ ]:
df.columns

# Salvando a base de dados

De forma correspondente, os Pandas nos permitem salvar o conjunto de dados csv pelo método dataframe.to_csv (), você pode adicionar o caminho e o nome do arquivo junto com as aspas no colchete.

Por exemplo, se você salvasse o dataframe "df" como "automobile.csv" em sua máquina local, você pode usar a sintaxe abaixo:

In [44]:
df.to_csv("automoveis.csv")

## Leitura e escrita Read/Save Outros formatos
Da mesma forma que lemos e salvamos arquivos csv, nós podemos usar métodos similares para ler e salvar outras bases em outros formatos:


| Formato dos dados | Leitura           | Salvar             |
| ------------- |:--------------:| ----------------:|
| csv           | `pd.read_csv()`  |`df.to_csv()`     |
| json          | `pd.read_json()` |`df.to_json()`    |
| excel         | `pd.read_excel()`|`df.to_excel()`   |
| hdf           | `pd.read_hdf()`  |`df.to_hdf()`     |
| sql           | `pd.read_sql()`  |`df.to_sql()`     |
| ...           |   ...          |       ...        |

<a id="ref2"></a>
# Insigths Básicos de um Banco de dados 

Depois de ler dados no quadro de dados da Pandas, é hora de explorar um pouco o conjunto de dados.

Existem várias maneiras de obter informações essenciais do conjunto de dados, para nos ajudar a entender melhor nosso conjunto de dados.



## Tipos de dados

Os dados têm uma variedade de tipos. Os principais tipos armazenados em objetos de pandas são `object`, `float`, `int`, `bool` e `datetime64`.Para melhor aprender sobre cada atributos, sempre é bom para nós conhecer o tipo de dados de cada coluna. Em Pandas:

~~~~
dataframe.dtypes
~~~~
retorna uma série com o tipo de dados de cada coluna.n.

In [45]:
# verifique o tipo de dados do quadro de dados "df" by .dtypes
df.dtypes

symboling              int64
normalized-losses     object
make                  object
fuel-type             object
aspiration            object
num-of-doors          object
body-style            object
drive-wheels          object
engine-location       object
wheel-base           float64
length               float64
width                float64
height               float64
curb-weight            int64
engine-type           object
num-of-cylinders      object
engine-size            int64
fuel-system           object
bore                  object
stroke                object
compression-ratio    float64
horsepower            object
peak-rpm              object
city-mpg               int64
highway-mpg            int64
price                 object
dtype: object


Como resultado mostrado acima, é claro que o tipo de dados de "symboling" e "curb-weight" são int64, "normalized-losses" são objetos e "wheel-base" é float64, etc. Na verdade, aqueles Os tipos de dados podem ser alterados de propósito, nós aprenderemos isso em um módulo posterior.


## Describe
Se quisermos verificar o resumo estatístico de cada coluna, como contagem de registros, valor médio da coluna, desvio padrão da coluna, etc.
~~~~
dataframe.describe()
~~~~
Gera várias estatísticas de resumo, excluindo `NaN` (Não é um número) valores.

In [49]:
# use .describe para obter o resumo estatístico de "df"
df.describe()

symboling  wheel-base  ...    city-mpg  highway-mpg
count  205.000000  205.000000  ...  205.000000   205.000000
mean     0.834146   98.756585  ...   25.219512    30.751220
std      1.245307    6.021776  ...    6.542142     6.886443
min     -2.000000   86.600000  ...   13.000000    16.000000
25%      0.000000   94.500000  ...   19.000000    25.000000
50%      1.000000   97.000000  ...   24.000000    30.000000
75%      2.000000  102.400000  ...   30.000000    34.000000
max      3.000000  120.900000  ...   49.000000    54.000000

[8 rows x 10 columns]



Aqui, ele mostra o resumo estatístico de todas as colunas numéricas (int, float).
Por exemplo, o atributo "simbolização" tem 205 contagens, o valor médio desta coluna é 0.83, o desvio padrão é 1.25, o valor mínimo é -2, o percentil 25 é 0, o percentil 50 é 1, o percentil 75 é 2 e o valor máximo é 3.
No entanto, e se nós também possamos verificar todas as colunas, incluindo outros tipos (como o objeto) de dados?
No entanto, e se nós também possamos verificar todas as colunas, incluindo outros tipos (como "objeto") de dados?

Você pode adicionar um argumento `include =" all "` dentro do suporte. Vamos tentar novamente.

In [50]:
# descreva todas as colunas em "df" 
df.describe(include = "all")

symboling normalized-losses    make  ...    city-mpg highway-mpg price
count   205.000000               205     205  ...  205.000000  205.000000   205
unique         NaN                52      22  ...         NaN         NaN   187
top            NaN                 ?  toyota  ...         NaN         NaN     ?
freq           NaN                41      32  ...         NaN         NaN     4
mean      0.834146               NaN     NaN  ...   25.219512   30.751220   NaN
std       1.245307               NaN     NaN  ...    6.542142    6.886443   NaN
min      -2.000000               NaN     NaN  ...   13.000000   16.000000   NaN
25%       0.000000               NaN     NaN  ...   19.000000   25.000000   NaN
50%       1.000000               NaN     NaN  ...   24.000000   30.000000   NaN
75%       2.000000               NaN     NaN  ...   30.000000   34.000000   NaN
max       3.000000               NaN     NaN  ...   49.000000   54.000000   NaN

[11 rows x 26 columns]


Agora, fornece o resumo estatístico de todas as colunas, incluindo atributos digitados por objetos. Agora podemos ver quantos valores únicos, que é o valor superior e a freqüência do maior valor em colunas com caracteres de objeto. Alguns valores na tabela acima mostram como "NaN", porque esse número não está disponível em relação ao tipo de coluna particular.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Questão #3: </h1>

<p>Você pode selecionar a coluna de um dataframe pela indicaçào do nome de cada coluna, por exemplo você pode selecionar o nome das três colunas a seguir </p>
<p></p>

<p>dataframe[[' column 1 ',column 2', 'column 3'] ]</p>
<p></p>

<p>onde "colums" é o nome da coluna, você pode aplicar o método ".describe()" para extrair as estatisticas das colunas como abaixo:</p>
<p></p>

<p>dataframe[[' column 1 ',column 2', 'column 3'] ].describe() </p>
<p></p>


Aplique o comando  ".describe()" para as colunas 'length' e 'compression-ratio'.
</div>

## Informações

Outro método que você pode usar para checar o seu dataset é:
~~~~
dataframe.info
~~~~
ele fornece um resumo conciso do seu dataframe.

In [51]:
# Veja a saída do com o comando .info "df"
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non

Aqui, podemos ver as informações do nosso quadro de dados, com as 30 primeiras linhas e as 30 linhas inferiores.

E também nos mostra que o quadro de dados inteiro tem 205 linhas e 26 colunas no total.

## Substituindo casos faltantes

In [53]:
df.dropna(subset=["price"], axis=0)

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0            3                 ?  alfa-romero  ...       21          27  13495
1            3                 ?  alfa-romero  ...       21          27  16500
2            1                 ?  alfa-romero  ...       19          26  16500
3            2               164         audi  ...       24          30  13950
4            2               164         audi  ...       18          22  17450
..         ...               ...          ...  ...      ...         ...    ...
200         -1                95        volvo  ...       23          28  16845
201         -1                95        volvo  ...       19          25  19045
202         -1                95        volvo  ...       18          23  21485
203         -1                95        volvo  ...       26          27  22470
204         -1                95        volvo  ...       19          25  22625

[205 rows x 26 columns]

In [56]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   symboling          205 non-null    int64  
 1   normalized-losses  205 non-null    object 
 2   make               205 non-null    object 
 3   fuel-type          205 non-null    object 
 4   aspiration         205 non-null    object 
 5   num-of-doors       205 non-null    object 
 6   body-style         205 non-null    object 
 7   drive-wheels       205 non-null    object 
 8   engine-location    205 non-null    object 
 9   wheel-base         205 non-null    float64
 10  length             205 non-null    float64
 11  width              205 non-null    float64
 12  height             205 non-null    float64
 13  curb-weight        205 non-null    int64  
 14  engine-type        205 non-null    object 
 15  num-of-cylinders   205 non-null    object 
 16  engine-size        205 non